File: python-algorithm/ds_5_tree/tree_4_build_tree/L105-m.cpp

#include <iostream>
#include <vector>
using namespace std ;

struct  TreeNode {
    int val ;
    TreeNode *left ;
    TreeNode *right ;
    TreeNode(int x) : val(x) , left(NULL) , right(NULL) {}
};

//思想没有问题，就是4个数组 效率比较低
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int inorder_size = inorder.size();
        if(inorder_size == 0)
            return NULL;
        vector<int> preorder_left, preorder_right, inorder_left, inorder_right;
        //创建根节点
        TreeNode* head = new TreeNode(preorder[0]);
        //找到中序遍历根节点所在位置,存放于变量gen中
        int gen = 0;
        for(int i = 0 ;i < inorder_size ; i++){
            if(inorder[i] == preorder[0]){
                gen = i;
                break;
            }
        }
        // 左子树
        for(int i = 0; i < gen; i++){
            inorder_left.push_back(inorder[i]);
            preorder_left.push_back(preorder[i+1]);//先序第一个为根节点
        }
        // 右子树
        for(int i = gen + 1; i < inorder_size; i++){
            inorder_right.push_back(inorder[i]);
            preorder_right.push_back(preorder[i]);
        }
        //递归，执行上述步骤，区分子树的左、右子子树，直到叶节点
        head->left = buildTree(preorder_left, inorder_left);
        head->right = buildTree(preorder_right, inorder_right);

        return head;
    }
};

//传索引的方式 效率比上面快10倍
#include <unordered_map>
class Solution1 {
private:
    unordered_map<int, int> um;
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int sz = preorder.size();
        // 构造哈希映射，帮助我们快速在中序遍历inorder中定位根节点，空间换时间的优化思想
        for(int i=0; i<sz; ++i)
            um[inorder[i]] = i;
        return myBuildTree(preorder, inorder, 0, sz-1, 0 , sz-1);
    }

    TreeNode* myBuildTree(vector<int>& preorder, vector<int>& inorder, int preorder_left,
                          int preorder_right, int inorder_left, int inorder_right){
        //递归终止条件
        if (preorder_left > preorder_right)
            return nullptr;
        // 前序遍历中的第一个节点就是根节点
        int preorder_root = preorder_left;
        // 在中序遍历中定位根节点
        int inorder_root = um[preorder[preorder_root]];
        // 先把根节点建立出来
        TreeNode* root = new TreeNode(preorder[preorder_root]);
        // 得到左子树中的节点数目
        int size_left_subtree = inorder_root - inorder_left;
        // 先序遍历中[preorder_left+1, preorder_left+size_left_subtree]的元素对应中序遍历[inorder_left, inorder_root-1]的元素
        root->left = myBuildTree(preorder, inorder, preorder_left+1, preorder_left+size_left_subtree, inorder_left, inorder_root-1);
        // 先序遍历中[preorder_left+size_left_subtree+1, preorder_right]的元素对应中序遍历[inorder_root + 1, inorder_right]的元素
        root->right = myBuildTree(preorder, inorder, preorder_left+size_left_subtree+1, preorder_right, inorder_root+1, inorder_right);
        return root;

    }
};

File: python-algorithm/ds_5_tree/tree_4_build_tree/L105-m.py
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# 递归不停的拆分左右子树
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        inorder_root_idxs = {}
        for i,n in enumerate(inorder):
            inorder_root_idxs[n] = i
        return self.build(preorder, inorder, 0, len(preorder)-1, 0, len(inorder)-1, inorder_root_idxs)

    def build(self, pro, ino, p_lt, p_rt, i_lt, i_rt, root_idx_dic):
        if p_lt>p_rt: return
        root_val = pro[p_lt]
        ino_root_idx = root_idx_dic[root_val]
        nums_of_left = ino_root_idx-i_lt
        root = TreeNode(root_val)
        root.left = self.build(pro, ino, p_lt+1, p_lt+nums_of_left, i_lt, ino_root_idx-1, root_idx_dic)
        root.right = self.build(pro, ino, p_lt+nums_of_left+1, p_rt, ino_root_idx+1, i_rt, root_idx_dic)
        return root